core/vm/actor/builtin/types/type_manager/type_manager.hpp

```diff
@@ -37,6 +40,12 @@ namespace fc::vm::actor::builtin::types {
         const std::vector<Multiaddress> &multiaddrs,
         const RegisteredSealProof &seal_proof_type,
         const RegisteredPoStProof &window_post_proof_type);
+
+    static outcome::result<Universal<Deadline>> makeEmptyDeadline(
+        const Runtime &runtime, const CID &empty_amt_cid);
+
+    static outcome::result<Deadlines> makeEmptyDeadlines(
+        const Runtime &runtime, const CID &empty_amt_cid);
```

Comment:
`Runtime` is unnecesssary, `IpldPtr` is sufficient (it contains actor version).

Convert static class method to method in namespace.

Move to "deadline.hpp" from "type_manager.hpp".

Seems that {v0,v2} reuse `empty_amy_cid` for all amt's in state (amt<v3 don't store bits), but {v3,v4,v5} make three insertions to `ipld` because amt bits differ.

Reply:
> `Runtime` is unnecesssary, `IpldPtr` is sufficient (it contains actor version).

We always get ipld from runtime. In general, you suggest to change `makeEmptyDeadlines(runtime)` to `makeEmptyDeadlines(runtime.getIpfsDatastore()`. Code with runtime looks more clean and easy.

> Convert static class method to method in namespace.

Probably it makes sense. I plan to redesign some common things (include TypeManager) in my next task FIL-379.

> Move to "deadline.hpp" from "type_manager.hpp".

It was design intent to move all factory methods to TypeManager. As I wrote above, I will consider it in FIL-379.
 
> Seems that {v0,v2} reuse `empty_amy_cid` for all amt's in state (amt<v3 don't store bits), but {v3,v4,v5} make three insertions to `ipld` because amt bits differ.

Yes

Reply:
I suggest to leave it as is. I think I will rework it in FIL-379